board: hsdk: add cy8c95xx I/O expander support

IRISZZW opened the pull request:
hsdk has an on board cy8c95xx I/O expander, some on board peripherals using expander GPIO.
This PR add initial support for cy8c95xx I/O expander, no interrupt support currently, then add
LEDs, arduino_header and arduino_header to hsdk dts, and remove ili9340 from hsdk dts, as it's not 
a part of hsdk board, it's a shield device.

IRISZZW committed:
drivers: gpio: add initial support for cy8c95xx I/O expander

add initial support for cy8c95xx I/O expander,
no interrupt support currently.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

IRISZZW committed:
boards: hsdk: dts: remove ili9340 from hsdk dts

remove ili9340 from hsdk dts, as it's not a part of hsdk board,
it's a shield device.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

IRISZZW committed:
boards: hsdk: add cy8c95xx I/O expander, LEDs support

hsdk has an on board cy8c95xx I/O expander, and 4 on
board LEDs use the expander GPIO. Add the I/O expander
and LEDs in hsdk dts, then add documents for them.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

IRISZZW committed:
boards: hsdk: add arduino_header and arduino_spi

add arduino_header and arduino_spi for hsdk baord.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

IRISZZW commented:
with this PR, `samples/basic/blinky` can run on hsdk board!

ruuddw commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,347 @@
+		      gpio_port_value_t toggle)
+{
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
```
Why do this before the local vars declaration? Other functions in this module do it after.

ruuddw commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,347 @@
+	}
+
+	/* Does not support pull-up/pull-down */
+	if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) != 0U) {
```
From the init code later, it looks like this IO expander does support pull up/down. Why not include it right away in the driver?

IRISZZW replied:
it's not support to configure driver mode for a single pin, it's only support PORT configuration.

mnkp replied:
According to the [documentation](https://www.cypress.com/file/372731/download)

>Drive Mode Registers (1Dh–23h)Each port’s data pins can be set separately to one of sevenavailable modes: pull up or down, open drain high/low, strongdrive  fast/slow,  or  high-impedance  input.  To  perform  thisconfiguration, the seven drive mode registers are used for theGPort selected by the Port Select register (18h). Each ‘1’ writtento  this  register  changes  the  corresponding  line  drive  mode.

it should be possible to configure pull up/down per pin. That said, I'm fine if you prefer not to implement and test all the features. If that's the case let's update the comments to state the feature is supported by the hardware but not implemented in the driver.

ruuddw commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,347 @@
+		    ((pin_data[2] << 12) & 0xF000);
+
+	if (rc != 0) {
+		goto out;
```
this goto and the out: label seem a bit overdone here, why not do it the same as in port_write?

ruuddw commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,347 @@
+	} else {
+		LOG_INF("%s init ok", dev->name);
+	}
+	k_sem_give(&drv_data->lock);
```
was there a k_sem_take before?

IRISZZW replied:
the sem's initial value will is 0, so we don't need to take it.

ruuddw reviewed (changes requested):
Please indicate in the PR title that it removes the display support, or even better do that in a different PR.

HSDK board documentation should be updated to reflect the changes, and maybe describe how the GPIOs can be used/tested.

Few comments in the code as well.

mnkp commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,329 @@
+	/* Does not support disconnected pin, and
+	 * not supporting both input/output at same time.
+	 */
+	io_flags = flags & (GPIO_INPUT | GPIO_OUTPUT);
+	if ((io_flags == GPIO_DISCONNECTED)
+	    || (io_flags == (GPIO_INPUT | GPIO_OUTPUT))) {
+		return -ENOTSUP;
+	}
```
According to the documentation the chip supports simultaneous read/write, Hi-Z pin state. It would be nice to have at least in/out mode supported as it makes it easier to test the driver.

IRISZZW replied:
I have some misunderstanding, it can be configured pull up/down per pin, I have added and tested. it works.

it also supports simultaneous read/write, when configured as output, I can read the pin and get the right value.

ruuddw replied:
I believe the device supports some more configurable features as well, like drive strength and open drain. Please either add the support (doesn't seem too complex), or mention in the comments that the driver doesn't (yet) support these features of the device.

mnkp commented on dts/bindings/gpio/cypress,cy8c95xx-gpio.yaml:
```diff
@@ -0,0 +1,40 @@
+      required: true
+      const: 16
+
+    init-dir:
+      type: int
+      required: false
+      default: 0
+      description: |
+        Bit mask identifying pins initial dir.
+
+    init-data:
+      type: int
+      required: false
+      default: 0
+      description: |
+        Bit mask identifying pins initial data.
```
This should be implemented using the so called [gpio hogging mechanism](https://github.com/torvalds/linux/blob/89698becf06d341a700913c3d89ce2a914af69a2/Documentation/devicetree/bindings/gpio/gpio.txt#L185) and respective properties. We don't support it currently. Unless you really need it I propose to remove above properties for now.

IRISZZW replied:
removed.

mnkp reviewed (commented):
This is a partial review.

IRISZZW commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,347 @@
+struct cy8c95xx_pin_state {
+	uint16_t dir;
+	uint16_t data;
+} __packed;
```
removed.

ruuddw commented on boards/arc/hsdk/doc/index.rst:
```diff
@@ -298,6 +298,56 @@ using the CREG_GPIO_MUX register (see Pinmux ). After a reset, all ports are con
+The ARC HSDK board includes a CY8C9520A I/O expander from `Cypress CY8C9520A
+<https://www.cypress.com/file/37971/download>`__. The I/O
+expander offers additional GPIO signals and board control signals and can be accessed
+through the on-board I2C bus, we have implemented a basice driver for it.
```
typo: basic, not basice

IRISZZW replied:
fixed.

galak commented:
dev-review (April 15, 2021): How to structure the devicetree for this.  Suggestion is to have a node per GPIO port as the device has up to 8 GPIO ports.

IRISZZW commented:
@mnkp @galak  Could you please give some advice how to structure the device tree to have a node per GPIO port. I have checked there is no similar design in zephyr currently. 
`gpio_pca95xx.c` has multi ports, but only write driver for one port: https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/gpio/gpio_pca95xx.c

is this right? :

```
&i2c0 {
	status = "okay";
	clock-frequency = <I2C_BITRATE_STANDARD>;
	cy8c95xx_port0: cy8c95xx_port0@20 {
		compatible = "cypress,cy8c95xx";
		reg = <0x20 0x00>;
		label = "cy8c95xx_io_port0";
		gpio-controller;
		#gpio-cells = <2>;
		ngpios = <8>;
	};
	cy8c95xx_port1: cy8c95xx_port1@20 {
		compatible = "cypress,cy8c95xx";
		reg = <0x20 0x01>;
		label = "cy8c95xx_io_port1";
		gpio-controller;
		#gpio-cells = <2>;
		ngpios = <8>;
	};
	// ...
	// ...
	cy8c95xx_port7: cy8c95xx_port7@20 {
		compatible = "cypress,cy8c95xx";
		reg = <0x20 0x07>;
		label = "cy8c95xx_io_port7";
		gpio-controller;
		#gpio-cells = <2>;
		ngpios = <8>;
	};
};
```

mnkp commented:
I believe it would be better to make it look similar to
- `dts/bindings/gpio/silabs,gecko-gpio.yaml`
- `dts/bindings/gpio/silabs,gecko-gpio-port.yaml`

The SiLabs GPIO module is a SoC internal one, but we could take a similar approach and split the device in one node with `compatible = "cypress,cy8c95xx"` and several nodes with `"compatible = "cypress,cy8c95xx-port"`. The reason for splitting is the architecture of the cy8c95xx. All the device registers are accessible via a single I2C address, this will be represented by `compatible = "cypress,cy8c95xx"` node. And the device can have up to 8 ports, they will be represented by "compatible = "cypress,cy8c95xx-port"`.

In all cases the `reg` property should have only address cell, no size cell, i.e. for `compatible = "cypress,cy8c95xx"` it should be `reg = <0x20>` representing the address on the I2C bus. For `"compatible = "cypress,cy8c95xx-port"` the reg property could represent the port number, i.e. be defined as a `reg = <n>` where n is a number between 0 and 7.

@galak, could you double check that the above approach is the right one.

IRISZZW commented:
> I believe it would be better to make it look similar to
> 
> * `dts/bindings/gpio/silabs,gecko-gpio.yaml`
> * `dts/bindings/gpio/silabs,gecko-gpio-port.yaml`
> 
> The SiLabs GPIO module is a SoC internal one, but we could take a similar approach and split the device in one node with `compatible = "cypress,cy8c95xx"` and several nodes with `"compatible = "cypress,cy8c95xx-port"`. The reason for splitting is the architecture of the cy8c95xx. All the device registers are accessible via a single I2C address, this will be represented by `compatible = "cypress,cy8c95xx"` node. And the device can have up to 8 ports, they will be represented by "compatible = "cypress,cy8c95xx-port"`.
> 
> In all cases the `reg` property should have only address cell, no size cell, i.e. for `compatible = "cypress,cy8c95xx"` it should be `reg = <0x20>` representing the address on the I2C bus. For `"compatible = "cypress,cy8c95xx-port"` the reg property could represent the port number, i.e. be defined as a `reg = <n>` where n is a number between 0 and 7.
> 
> @galak, could you double check that the above approach is the right one.

Thanks @mnkp , I have updated the codes, Can you help review again.

mnkp commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,304 @@
+/** Cache of the output configuration and data of the pins. */
+struct cy8c95xx_pin_state {
+	uint8_t dir;
+	uint8_t data;
```
It might be worth to rename the field to `data_out`, `out_data` or similar to make it easier to notice that this holds output port values (in contrary to input port values).

mnkp commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,304 @@
+struct cy8c95xx_drv_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	const struct device *i2c_master;
```
@galak should we require that this is moved to the const driver config and resolved at build time via `DEVICE_DT_GET`?

mnkp replied:
It looks like we should use the new `DEVICE_DT_GET` macro to get handles of device driver dependencies. The macro allows us to get the handle at compile time. It's a more robust approach and has smaller footprint. This field can be moved to `struct cy8c95xx_config`. The `const char *i2c_master_dev_name;` in the other struct can be removed. There will also be no more need to call `device_get_binding` in the init function.

You can grep for `DEVICE_DT_GET` in `drivers/` folder for usage examples.

mnkp commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,304 @@
+{
+	int rc;
+
+	rc = i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				CY8C95XX_REG_PORT_SELECT, cfg->port_num);
+	if (rc == 0) {
+		i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				   CY8C95XX_REG_DIR, pins->dir);
+		i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				   CY8C95XX_REG_PULL_UP, pins->pullup);
+		i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				   CY8C95XX_REG_PULL_DOWN, ~pins->pullup);
+	}
+	rc = i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				CY8C95XX_REG_OUTPUT_DATA0 + cfg->port_num, pins->data);
```
We should write to the output data register before we write to the pin direction register. If not, user may observe a glitch on the output.

We should verify return value of every call to `i2c_reg_write_byte`.

mnkp commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,304 @@
+		return -ENOTSUP;
+	}
+
+	/* No open-drain support */
```
Let's change the comment to clarify it's not implemented by the driver. The feature is supported by hardware. Similar change should be done for drive strength, pin disconnected.

mnkp commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,304 @@
+	} else {
+		pins->pullup &= ~BIT(pin);
+	}
+
```
Let's implement handling of `GPIO_PULL_DOWN` flag or return `-ENOTSUP` if flag is set.

IRISZZW replied:
I have implemented GPIO_PULL_DOWN, if you don't set GPIO_PULL_UP, then it's GPIO_PULL_DOWN.

mnkp replied:
Sorry, I missed that. Nevertheless, it's incorrect. Pull up, pull down have to be configured independently. Missing pull up flag does not mean that user requests pull down.

mnkp replied:
This fragment should be written as
```
	if ((flags & GPIO_PULL_UP) != 0) {
		pins->pullup |= BIT(pin);
	} else {
		pins->pullup &= ~BIT(pin);
	}

	if ((flags & GPIO_PULL_DOWN) != 0) {
		pins->pulldown |= BIT(pin);
	} else {
		pins->pulldown &= ~BIT(pin);
	}
```
or preferably in a more concise manner using `WRITE_BIT` macro
```
	WRITE_BIT(pins->pullup, pin, (flags & GPIO_PULL_UP) != 0U);
	WRITE_BIT(pins->pulldown, pin, (flags & GPIO_PULL_DOWN) != 0U);
```
You'll need to add `pulldown` field to `struct cy8c95xx_pin_state`.

mnkp commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,304 @@
+		return -ENOTSUP;
+	}
+
+	/* Default: pull down */
```
The comment is a misleading, maybe we should remove it?

mnkp commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,304 @@
+	drv_data->pin_state = (struct cy8c95xx_pin_state) {
+		.dir = 0xFF,
+		.data = 0x00,
+		.pullup = 0x00,
```
According to the datasheet the hardware is initialized with pins configured in Resistive Pull Up mode. We shouldn't change it or we may break some boards. Board hardware designer will rely on the Resistive Pull Up mode selected and all pins being driven high. Pin state should change only when explicitly requested by calling GPIO API. It might be even better to read the initial state from the hardware registers rather than write it.

IRISZZW replied:
I will set pull up mode as the default mode.

mnkp reviewed (commented):
LGTM with some minor comments.

I believe we should squash the following two commits, since they describe one logical change
- boards: hsdk: add documents for I/O expander and LEDs
- boards: hsdk: add cy8c95xx I/O expander, LEDs support

Let's test the driver with `tests/drivers/gpio/gpio_api_1pin/` testcase.

mnkp commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,317 @@
+	/* Reset state mediated by initial configuration */
+	drv_data->pin_state = (struct cy8c95xx_pin_state) {
+		.dir = 0xFF,
+		.data_out = 0x00,
```
We also need to keep `data_out = 0xFF`, since that's the default power up value. The `write_pin_state` writes to the data output register first, pin direction register later. With `data_out = 0x00` we would observe glitch on the outputs during device init.

mnkp reviewed (commented):
Thanks, that looks good. Just the last few bits.

Let's squash the following two commits, since they describe one logical change
- boards: hsdk: add documents for I/O expander and LEDs
- boards: hsdk: add cy8c95xx I/O expander, LEDs support

IRISZZW commented:
> Thanks, that looks good. Just the last few bits.
> 
> Let's squash the following two commits, since they describe one logical change
> 
> * boards: hsdk: add documents for I/O expander and LEDs
> * boards: hsdk: add cy8c95xx I/O expander, LEDs support

thanks for your review, fixed.

IRISZZW commented:
`tests/drivers/gpio/gpio_api_1pin` results:

```
*** Booting Zephyr OS build zephyr-v2.5.0-2554-gaa813e22470c  ***
Running test suite gpio_api_1pin_test
===================================================================
START - test_gpio_pin_configure_push_pull
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_pin_configure_push_pull in 0.108 seconds
===================================================================
START - test_gpio_pin_configure_single_ended
Running test on port=cy8c95xx_port1, pin=4
When configured as input test pin value is low
Open Source configuration or Pull Down pin bias is not supported
 SKIP - test_gpio_pin_configure_single_ended in 0.136 seconds
===================================================================
START - test_gpio_pin_set_get_raw
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_pin_set_get_raw in 0.62 seconds
===================================================================
START - test_gpio_pin_set_get
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_pin_set_get in 0.62 seconds
===================================================================
START - test_gpio_pin_set_get_active_high
Running test on port=cy8c95xx_port1, pin=4
Step 1: Set logical, get logical and physical pin value
Step 2: Set physical, get logical and physical pin value
 PASS - test_gpio_pin_set_get_active_high in 0.144 seconds
===================================================================
START - test_gpio_pin_set_get_active_low
Running test on port=cy8c95xx_port1, pin=4
Step 1: Set logical, get logical and physical pin value
Step 2: Set physical, get logical and physical pin value
 PASS - test_gpio_pin_set_get_active_low in 0.132 seconds
===================================================================
START - test_gpio_pin_toggle
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_pin_toggle in 0.36 seconds
===================================================================
START - test_gpio_port_set_masked_get_raw
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_port_set_masked_get_raw in 0.65 seconds
===================================================================
START - test_gpio_port_set_masked_get
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_port_set_masked_get in 0.66 seconds
===================================================================
START - test_gpio_port_set_masked_get_active_high
Running test on port=cy8c95xx_port1, pin=4
Step 1: Set logical, get logical and physical port value
Step 2: Set physical, get logical and physical port value
 PASS - test_gpio_port_set_masked_get_active_high in 0.192 seconds
===================================================================
START - test_gpio_port_set_masked_get_active_low
Running test on port=cy8c95xx_port1, pin=4
Step 1: Set logical, get logical and physical port value
Step 2: Set physical, get logical and physical port value
 PASS - test_gpio_port_set_masked_get_active_low in 0.192 seconds
===================================================================
START - test_gpio_port_set_bits_clear_bits_raw
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_port_set_bits_clear_bits_raw in 0.54 seconds
===================================================================
START - test_gpio_port_set_bits_clear_bits
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_port_set_bits_clear_bits in 0.53 seconds
===================================================================
START - test_gpio_port_set_clr_bits_raw
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_port_set_clr_bits_raw in 0.42 seconds
===================================================================
START - test_gpio_port_set_clr_bits
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_port_set_clr_bits in 0.38 seconds
===================================================================
START - test_gpio_port_toggle
Running test on port=cy8c95xx_port1, pin=4
 PASS - test_gpio_port_toggle in 0.36 seconds
===================================================================
START - test_gpio_int_edge_rising
Running test on port=cy8c95xx_port1, pin=4
Pin interrupt is not supported.
 SKIP - test_gpio_int_edge_rising in 0.27 seconds
===================================================================
START - test_gpio_int_edge_falling
Running test on port=cy8c95xx_port1, pin=4
Pin interrupt is not supported.
 SKIP - test_gpio_int_edge_falling in 0.27 seconds
===================================================================
START - test_gpio_int_edge_both
Running test on port=cy8c95xx_port1, pin=4
Pin interrupt is not supported.
 SKIP - test_gpio_int_edge_both in 0.27 seconds
===================================================================
START - test_gpio_int_edge_to_active
Step 1: Configure pin as active high
Running test on port=cy8c95xx_port1, pin=4
Pin interrupt is not supported.
 SKIP - test_gpio_int_edge_to_active in 0.30 seconds
===================================================================
START - test_gpio_int_edge_to_inactive
Step 1: Configure pin as active high
Running test on port=cy8c95xx_port1, pin=4
Pin interrupt is not supported.
 SKIP - test_gpio_int_edge_to_inactive in 0.30 seconds
===================================================================
START - test_gpio_int_level_high
Running test on port=cy8c95xx_port1, pin=4
Pin interrupt is not supported.
 SKIP - test_gpio_int_level_high in 0.27 seconds
===================================================================
START - test_gpio_int_level_low
Running test on port=cy8c95xx_port1, pin=4
Pin interrupt is not supported.
 SKIP - test_gpio_int_level_low in 0.27 seconds
===================================================================
START - test_gpio_int_level_active
Step 1: Configure pin as active high
Running test on port=cy8c95xx_port1, pin=4
Pin interrupt is not supported.
 SKIP - test_gpio_int_level_active in 0.30 seconds
===================================================================
START - test_gpio_int_level_inactive
Step 1: Configure pin as active high
Running test on port=cy8c95xx_port1, pin=4
Pin interrupt is not supported.
 SKIP - test_gpio_int_level_inactive in 0.30 seconds
===================================================================
START - test_gpio_pin_toggle_visual
Running test on port=cy8c95xx_port1, pin=4
LED ON
LED OFF
LED ON
LED OFF
 PASS - test_gpio_pin_toggle_visual in 6.50 seconds
===================================================================
Test suite gpio_api_1pin_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

mnkp commented on drivers/gpio/gpio_cy8c95xx.c:
```diff
@@ -0,0 +1,314 @@
+{
+	int rc;
+
+	rc = i2c_reg_write_byte(cfg->i2c_master, cfg->i2c_slave_addr,
+				CY8C95XX_REG_OUTPUT_DATA0 + cfg->port_num, pins->data_out);
+	if (rc) {
+		return rc;
+	}
+
+	rc = i2c_reg_write_byte(cfg->i2c_master, cfg->i2c_slave_addr,
+				CY8C95XX_REG_PORT_SELECT, cfg->port_num);
+	if (rc) {
+		return rc;
+	}
```
According to the datasheet we need to write to Port Select register first, before writing to registers 19h–23h. So this two calls should probably be swapped. The testcase is passing since the port1 to which test pin is connected was the last one setup by the `cy8c95xx_init` function.

IRISZZW replied:
`#define CY8C95XX_REG_OUTPUT_DATA0 0x08`
each port has its own output data register, we don't need to write port select register first.

mnkp replied:
Sorry, my bad. Thanks!

mnkp reviewed (commented):
Looks very good. One last issue to fix.

ruuddw reviewed (approved):
Thanks Watson (and Piotr for the reviews), looks good now.
